onnxscript/rewriter/onnxruntime/bfloat16_utils/bfloat16_converter.py

```diff
@@ -0,0 +1,84 @@
+import logging
+
+from onnxscript import ir
+
+logger = logging.getLogger(__name__)
+
+
+def _convert_inputs_from_bfloat16_to_float16(value: ir.Input):
+    if value.dtype == ir.DataType.BFLOAT16:
+        value.dtype = ir.DataType.FLOAT16
+        value_users = list(value.uses())
+        for node, index in value_users:
+            _insert_cast_nodes_for_float16_to_bfloat16_to_inputs(node, index)
+
+
+def _convert_outputs_from_bfloat16_to_float16(value: ir.Value):
+    if value.dtype == ir.DataType.BFLOAT16:
+        node = value.producer()
+        index = value.index()
+        if node is None or index is None:
+            logger.warning("Output value %s has no producer or index", value)
+            return
+        _insert_cast_nodes_for_bfloat16_to_float16_to_outputs(node, index)
+
+
+def _insert_cast_nodes_for_float16_to_bfloat16_to_inputs(node: ir.Node, index: int):
+    attr = ir.AttrInt64(name="to", value=ir.DataType.BFLOAT16)
+    cast = ir.Node(
+        domain=node.domain,
+        op_type="Cast",
+        inputs=[node.inputs[index]],
+        num_outputs=1,
+        attributes=[attr],
+    )
+    cast.outputs[0].dtype = ir.DataType.BFLOAT16  # type: ignore[union-attr]
+    cast.outputs[0].shape = node.inputs[index].shape  # type: ignore[union-attr]
+    node.prepend(cast)
+    node.replace_input_with(index, cast.outputs[0])
+
+
+def _insert_cast_nodes_for_bfloat16_to_float16_to_outputs(node: ir.Node, index: int):
+    attr = ir.AttrInt64(name="to", value=ir.DataType.FLOAT16)
+    cast = ir.Node(
+        domain=node.domain,
+        op_type="Cast",
+        inputs=[node.outputs[index]],
+        num_outputs=1,
+        attributes=[attr],
+    )
+    cast.outputs[0].dtype = ir.DataType.FLOAT16
+    cast.outputs[0].shape = node.outputs[index].shape
+    node.append(cast)
+    if node.graph is None:
+        logger.warning("Node %s has no graph", node)
+        return
+    # Update graph/function outputs
+    for idx, graph_or_function_output in enumerate(node.graph.outputs):
+        if graph_or_function_output == node.outputs[index]:
+            node.graph.outputs[idx] = cast.outputs[0]
+
+
+def dtype_adapter_for_bfloat16_model(model: ir.Model) -> ir.Model:
```

Comment:
```suggestion
def dtype_adapter_for_bfloat16_model(model: ir.Model) -> None:
```

Model is updated in-place so we don't need to return, for clarity.